test/Microsoft.Azure.Devices.Edge.Test.Common/YamlDocument.cs

```diff
@@ -41,6 +41,24 @@ public void ReplaceOrAdd(string dottedKey, string value)
             node[leaf] = value;
         }
 
+        public void RemoveIfExists(string dottedKey)
+        {
+            Dictionary<object, object> node = this.root;
+            string[] segments = dottedKey.Split('.');
+            foreach (string key in segments.SkipLast(1))
```

Comment:
Suggestion: extract this code block as a method "findNode"; which can be used in "ReplaceOrAdd" method.

Reply:
I tried to write `FindNode`, but decided I didn't like the extra processing code you have to write in the caller to deal with the returned node.

Since the add and remove functions only differ in two places--how to handle a non-terminal key segment that doesn't have a node in the document, and what to do with the terminal key segment after finding the terminal node--I wrote a function that encapsulates tree traversal and takes two lambdas. Check out my latest commits.

Reply:
Review with new version with 2 lambdas; I think it is more readable with previous version without code sharing.

Reply:
`
public void AddOrReplace(string dottedKey, string value)
            {
                Node node = this.root;
                string[] keys = dottedKey.Split('.');

                int i = 0;
                for (i = 0; i < keys.Length - 1; i++)
                {
                    if (!node.ContainsKey(keys[i]))
                    {
                        node = AddNodes(node, keys.TakeLast(keys.Length - i));
                        break;
                    }

                    node = (Node) node[keys[i]];
                }

                node[keys.Last()] = value;
            }

            private Node AddNodes(Node baseNode, IEnumerable<string> keys)
            {
                Node tempNode = baseNode;
                
                foreach (string key in keys)
                {
                    var newNode = new Node();
                    tempNode.Add(key, newNode);
                    tempNode = newNode;
                }

                return tempNode;
            }

            public void Remove(string dottedKey)
            {
                Node node = this.root;
                string[] keys = dottedKey.Split('.');

                int i = 0;
                for (i = 0; i < keys.Length - 1; i++)
                {
                    if (!node.ContainsKey(keys[i]))
                    {
                        return;
                    }

                    if (i != keys.Length)
                    {
                        continue;
                    }

                    node.Remove(keys[i]);
                }
            }
`

Reply:
> I think it is more readable with previous version without code sharing.

I agree, which is why I didn't originally share code between the two functions when I added `RemoveIfExists`. I'm happy to put it back the way it was.

Reply:
Reverted.